python/lsst/daf/butler/tests/hybrid_butler.py

```diff
@@ -0,0 +1,431 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This software is dual licensed under the GNU General Public License and also
+# under a 3-clause BSD license. Recipients may choose which of these licenses
+# to use; please see the files gpl-3.0.txt and/or bsd_license.txt,
+# respectively.  If you choose the GPL option then the following text applies
+# (but note that there is still no warranty even if you opt for BSD instead):
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from collections.abc import Collection, Iterable, Mapping, Sequence
+from contextlib import AbstractContextManager
+from typing import Any, TextIO, cast
+
+from lsst.resources import ResourcePath, ResourcePathExpression
+
+from .._butler import Butler
+from .._dataset_existence import DatasetExistence
+from .._dataset_ref import DatasetId, DatasetRef
+from .._dataset_type import DatasetType
+from .._deferredDatasetHandle import DeferredDatasetHandle
+from .._file_dataset import FileDataset
+from .._limited_butler import LimitedButler
+from .._query import Query
+from .._storage_class import StorageClass
+from .._timespan import Timespan
+from ..datastore import DatasetRefURIs
+from ..dimensions import DataCoordinate, DataId, DimensionGroup, DimensionRecord, DimensionUniverse
+from ..registry import CollectionArgType, Registry
+from ..remote_butler import RemoteButler
+from ..transfers import RepoExportContext
+
+
+class HybridButler(Butler):
+    """A `Butler` that delegates methods to internal RemoteButler and
+    DirectButler instances.  Intended to allow testing of RemoteButler before
+    its implementation is complete, by delegating unsupported methods to
+    DirectButler.
+    """
+
+    _remote_butler: RemoteButler
+    _direct_butler: Butler
+
+    def __new__(cls, remote_butler: RemoteButler, direct_butler: Butler) -> "HybridButler":
```

Comment:
```suggestion
    def __new__(cls, remote_butler: RemoteButler, direct_butler: Butler) -> Self:
```
Does `typing.Self` work here? (maybe also need `from __future__ import annotations` to remove the quotes?)

Reply:
Self doesn't for some reason... it causes mypy to throw "expected Self but got HybridButler".

`from __future__ import annotations` works though, thanks -- I could not for the life of me figure out why this exact same pattern worked in DirectButler and RemoteButler but not in this file.